flink-batch-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCInputFormat.java

```diff
@@ -139,12 +184,13 @@ public void close() throws IOException {
 	@Override
 	public boolean reachedEnd() throws IOException {
 		try {
-			if (resultSet.isLast()) {
-				close();
+			if (!hasNext || resultSet.isLast()) {
+				close();	
```

Comment:
I think `reachedEnd()` should not call `close()`. `close()` should be called by the `DataSourceTask` (or the other class handling the `InputFormat`)